RecoTracker/FinalTrackSelectors/plugins/TrackLwtnnClassifier.cc

```diff
@@ -0,0 +1,124 @@
+    }
+
+    void beginStream() {}
+    void initEvent(const edm::EventSetup& es) {
+      edm::ESHandle<lwt::LightweightNeuralNetwork> lwtnnHandle;
+      es.get<TrackingComponentsRecord>().get(lwtnnLabel_, lwtnnHandle);
+      neuralNetwork_ = lwtnnHandle.product();
+    }
+
+    float operator()(reco::Track const & trk,
+                     reco::BeamSpot const & beamSpot,
+                     reco::VertexCollection const & vertices,
+                     lwt::ValueMap & inputs) const {
+      // lwt::ValueMap is typedef for std::map<std::string, double>
+      //
+      // It is cached per event to avoid constructing the map for each
+      // track while keeping the operator() interface const.
+
+      Point bestVertex = getBestVertex(trk,vertices);
+
+      inputs["trk_pt"] = trk.pt();
+      inputs["trk_eta"] = trk.eta();
+      inputs["trk_lambda"] = trk.lambda();
+      inputs["trk_dxy"] = trk.dxy(beamSpot.position()); // Training done without taking absolute value
+      inputs["trk_dz"] = trk.dz(beamSpot.position()); // Training done without taking absolute value
+      inputs["trk_dxyClosestPV"] = trk.dxy(bestVertex); // Training done without taking absolute value
+      inputs["trk_dzClosestPVNorm"] = std::max(-0.2, std::min(trk.dz(bestVertex), 0.2)); // Training done without taking absolute value
+      inputs["trk_ptErr"] = trk.ptError();
+      inputs["trk_etaErr"] = trk.etaError();
+      inputs["trk_lambdaErr"] = trk.lambdaError();
+      inputs["trk_dxyErr"] = trk.dxyError();
+      inputs["trk_dzErr"] = trk.dzError();
+      inputs["trk_nChi2"] = trk.normalizedChi2();
+      inputs["trk_ndof"] = trk.ndof();
+      inputs["trk_nInvalid"] = trk.hitPattern().numberOfLostHits(reco::HitPattern::TRACK_HITS);
+      inputs["trk_nPixel"] = trk.hitPattern().numberOfValidPixelHits();
+      inputs["trk_nStrip"] = trk.hitPattern().numberOfValidStripHits();
+      inputs["trk_nPixelLay"] = trk.hitPattern().pixelLayersWithMeasurement();
+      inputs["trk_nStripLay"] = trk.hitPattern().stripLayersWithMeasurement();
+      inputs["trk_n3DLay"] = (trk.hitPattern().numberOfValidStripLayersWithMonoAndStereo()+trk.hitPattern().pixelLayersWithMeasurement());
+      inputs["trk_nLostLay"] = trk.hitPattern().trackerLayersWithoutMeasurement(reco::HitPattern::TRACK_HITS);
+      inputs["trk_algo"] = trk.algo(); // eventually move to originalAlgo
+
+      auto out = neuralNetwork_->compute(inputs);
+      // there should only one output
+      if(out.size() != 1) throw cms::Exception("LogicError") << "Expecting exactly one output from NN, got " << out.size();
+
+      float output = 2.0*out.begin()->second-1.0;
+
+      //Special clauses for special tracks
+      //T1qqqq
+      if((std::abs(inputs["trk_dxy"])>=0.1) && (inputs["trk_etaErr"]<0.003) && (inputs["trk_dxyErr"]<0.03) &&(inputs["trk_ndof"]>3)){
+        //Set value to just above the threshold
+        if(generalCollThresholds[trk.algo()]){
```

Comment:
what is the purpose of this ```if ```?
Current logic will skip
- cases where generalCollThresholds is not present
    - shouldn't this be an empty set in this context (because the NN is called, it's called for a reason to compute values and select)
- cases where the generalCollThresholds is present but it is set to 0
    - it looks like negative values of the score are possible for some iterations, OTOH, some iterations have the cut set to 0. 
    - Is there some magic in value 0 that the selection bypass is disallowed?

Reply:
That was an honest error on my part. I think the original purpose was to test if the dictionary contains the given key and handle the possible situation of key not existing in the dictionary somehow. However the implementation was clearly wrong and as you mentioned, there should't be a case where the key wouldn't exist since the NN being is called implies this being one of the implemented iterations.

This is no longer a problem in the new implementation.